logstage/logstage-core/src/main/scala-2/izumi/logstage/macros/LogMethodMacro.scala

```diff
@@ -0,0 +1,193 @@
+package izumi.logstage.macros
+
+import izumi.functional.quasi.{QuasiIO, QuasiPrimitives}
+import izumi.fundamentals.platform.language.CodePositionMaterializer.CodePositionMaterializerMacro.getEnclosingPosition
+import izumi.logstage.api.Log.Level
+
+import scala.annotation.tailrec
+import scala.reflect.macros.blackbox
+
+class LogMethodMacro[C <: blackbox.Context](final val c: C) {
+  import c.universe.*
+
+  def logMethodIOF[F[_], A](
+    level: c.Expr[Level],
+    function: c.Expr[F[A]],
+    logTypes: Boolean,
+    logImplicits: Boolean,
+    qp: c.Expr[QuasiPrimitives[F]],
+  ): c.Expr[F[A]] = {
+    val (variables, logString) = createVariablesAndLogStringTrees(function, logTypes, logImplicits)
+
+    val logTree =
+      q"""
+         $qp.tapBothUntyped($function)(
+           err = error => self.log($level)(_root_.izumi.logstage.api.Log.Message.apply($logString + " => " + error))(position),
+           succ = result => self.log($level)(_root_.izumi.logstage.api.Log.Message.apply($logString + " => " + result))(position)
+         )   
+        """
+
+    c.Expr[F[A]](q"""
+           val self = ${c.prefix}
+           val position = ${getEnclosingPosition(c)}
+           ..$variables
+           $logTree
+         """)
+  }
+
+  def logMethodIO[F[_], A](
+    level: c.Expr[Level],
+    function: c.Expr[A],
+    logTypes: Boolean,
+    logImplicits: Boolean,
+    qp: c.Expr[QuasiIO[F]],
+  ): c.Expr[F[A]] = {
+    val (variables, logString) = createVariablesAndLogStringTrees(function, logTypes, logImplicits)
```

Comment:
```suggestion
    logMethodIOF(level, c.Expr[F[A]](q"$qp.maybeSuspend($function)"), logTypes, logImplicits, qp)
```

Should work, since c.Expr[QuasiIO[F]] is <: c.Expr[QuasiPrimitives[F]]